pkg/controller/network/internetgateway/controller.go

```diff
@@ -23,205 +23,255 @@ import (
-	// categorize immutable vs mutable fields (see #727)
+	cr, ok := mgd.(*v1beta1.InternetGateway)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errUnexpectedObject)
+	}
+
+	if cr.Status.AtProvider.InternetGatewayID == "" {
+		return managed.ExternalUpdate{}, errors.New(errUpdatePresent)
+	}
+
+	response, err := e.client.DescribeInternetGatewaysRequest(&awsec2.DescribeInternetGatewaysInput{
+		InternetGatewayIds: []string{cr.Status.AtProvider.InternetGatewayID},
+	}).Send(ctx)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(resource.Ignore(ec2.IsInternetGatewayNotFoundErr, err), errDescribe)
+	}
+
+	observed := response.InternetGateways[0]
+
+	// Although, observed.Attachments is a list, an internet gateway can be attached to only one
+	// VPC. DetachInternetGatewayRequest returns error for an unattached VPCID.
+	// Detaching the currently attached VPC below.
+	var attached bool
+	if observed.Attachments != nil {
+		for _, attachment := range observed.Attachments {
+			if *attachment.VpcId == cr.Spec.ForProvider.VPCID {
+				attached = true
+				break
+			} else {
+				_, err := e.client.DetachInternetGatewayRequest(&awsec2.DetachInternetGatewayInput{
+					InternetGatewayId: aws.String(cr.Status.AtProvider.InternetGatewayID),
+					VpcId:             aws.String(*attachment.VpcId),
+				}).Send(ctx)
+
+				if err != nil {
+					return managed.ExternalUpdate{}, errors.Wrapf(err, errUpdate)
+				}
+			}
+		}
+	}
+
+	if !attached && cr.Spec.ForProvider.VPCID != "" {
+		// Attach IG to VPC in spec.
+		_, err = e.client.AttachInternetGatewayRequest(&awsec2.AttachInternetGatewayInput{
+			InternetGatewayId: aws.String(cr.Status.AtProvider.InternetGatewayID),
+			VpcId:             aws.String(cr.Spec.ForProvider.VPCID),
+		}).Send(ctx)
+	}
 
-	return managed.ExternalUpdate{}, nil
+	return managed.ExternalUpdate{}, errors.Wrap(err, errUpdate)
 }
 
 func (e *external) Delete(ctx context.Context, mgd resource.Managed) error {
-	cr, ok := mgd.(*v1alpha3.InternetGateway)
+	cr, ok := mgd.(*v1beta1.InternetGateway)
 	if !ok {
 		return errors.New(errUnexpectedObject)
 	}
 
-	if cr.Status.InternetGatewayID == "" {
+	if cr.Status.AtProvider.InternetGatewayID == "" {
 		return errors.New(errDeleteNotPresent)
 	}
 
 	cr.Status.SetConditions(runtimev1alpha1.Deleting())
 
 	// first detach all vpc attachments
-	for _, a := range cr.Status.Attachments {
-		// after creating the IG, attach the VPC
-		dReq := e.client.DetachInternetGatewayRequest(&awsec2.DetachInternetGatewayInput{
-			InternetGatewayId: aws.String(cr.Status.InternetGatewayID),
+	for _, a := range cr.Status.AtProvider.Attachments {
+		_, err := e.client.DetachInternetGatewayRequest(&awsec2.DetachInternetGatewayInput{
+			InternetGatewayId: aws.String(cr.Status.AtProvider.InternetGatewayID),
 			VpcId:             aws.String(a.VPCID),
-		})
+		}).Send(ctx)
 
-		if _, err := dReq.Send(ctx); err != nil {
+		if err != nil {
```

Comment:
Avoiding nested if statements is considered more idiomatic. So, maybe you can check for `NotFound` and then in a separate if statement you can check `if resource.Ignore(ec2.IsInternetGatewayNotFoundErr, err) != nil`